### ISSUE TITLE
Is it possible to compile cpgf to dll project?

### ISSUE THREAD

**Issue by cutepig123** [cpgf/cpgf#80]
I have multiple dlls. Each dll is a plugin. Every dll needs to register itself using cpgf so that their classes/functions can be reflected. 
If cpgf is static linked by each dll, then each dll should have a copy of cpgf, they cannot share a same cpgf instance
The only solution I know is to make cpgf as a dll. so that every plugin can share same cpgf instance. 
So my question is

- Is compile cpgf to dll project the only solution?
- Is it possible to compile cpgf to dll project?

**Comment by wqking** [comment]
There is LIB_SHARED in build.config.txt, you may try to enable it. It was added by another developer, I'm not sure if it works now.
Let me know if it works and we may do more investigation.

**Comment by cutepig123** [comment]
I do a rough checking by search codes. I think it does not work in windows platform..

```
Find all "declspec", Subfolders, Find Results 2, "C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1", "*"
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\include\cpgf\gdeprecated.h(7):	#define G_DEPRECATED(func, msg) __declspec(deprecated(msg)) func
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\include\cpgf\thirdparty\jsoncpp\config.h(47):#  define JSON_API __declspec(dllexport)
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\include\cpgf\thirdparty\jsoncpp\config.h(49):#  define JSON_API __declspec(dllimport)
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\include\cpgf\thirdparty\jsoncpp\config.h(66):# define JSONCPP_DEPRECATED(message) __declspec(deprecated(message))
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\test\catch.hpp(1387):    extern "C" __declspec(dllimport) void __stdcall DebugBreak();
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\test\catch.hpp(6288):    extern "C" __declspec(dllimport) int __stdcall IsDebuggerPresent();
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\test\catch.hpp(6295):    extern "C" __declspec(dllimport) int __stdcall IsDebuggerPresent();
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\test\benchmark\callback\benchmarkcallback.h(7):	#define NON_INLINE __declspec(noinline)
  C:\\Downloads\cpgf-1.6.1\cpgf-1.6.1\test\misc\test_callback.cpp(15):	#define NON_INLINE __declspec(noinline)
  Matching lines: 9    Matching files: 5    Total files searched: 1538

```

**Comment by wqking** [comment]
OK, I will work on the Windows version.

**Comment by wqking** [comment]
Each DLL is a module, each module has its meta data. So it's impossible (at least for now) that several modules register the meta data to a single repository.  
However, cpgf supports multiple modules via the interface based API. This is a [very brief (and fuzzy) document](http://www.cpgf.org/document/using-interface-api.html).  
In brief, you each plugin should export a function with whatever name such as `IMetaModule * getMetaModule()`, your application get the meta modules, and add them to a `IMetaService`. You can call `createDefaultMetaService()` to create a meta server from within your application. Then you can call `IMetaService::findClassByName` to find any meta class in any modules.

The serialization and script binding uses meta service extensively.

**Comment by cutepig123** [comment]
> 
> 
> @cutepig123 Just to be clear, what's your problem to solve? Seems the problem is not to compile cpgf to DLL. I need to understand the essential problem. And what you mean by "same cpgf instance"? I guess you want to use `getGlobalMetaClass()` or `findMetaClass` to get meta class in different plugins (DLLs)?

Yes. I have multiple dlls. I need these dlls can share the reflection data (i.e. the metaclass info)
Can interface based API help on this?

**Comment by wqking** [comment]
Yes. You can get all meta data from IMetaService. But of course you need to add all modules to IMetaService first.
